joint_teapot/utils/joj3.py

```diff
@@ -1,14 +1,139 @@
+            self.date = ""
+            self.repository = self.Link()
+            self.failure = self.Link()
+
+        def init(
+            self,
+            date: str,
+            repo_name: str,
+            repo_link: str,
+            failure_name: str,
+            failure_link: str,
+        ) -> Failed_Table.Row:
+            self.date = date
+            self.repository.init(repo_name, repo_link)
+            self.failure.init(failure_name, failure_link)
+            return self
+
+        def init_from_line(self, line: list[str]) -> Failed_Table.Row:
+            self.date = line[0]
+            self.repository.init_from_string(line[1])
+            self.failure.init_from_string(line[2])
+            return self
+
+        def __str__(self) -> str:
+            return f"|{self.date}|{str(self.repository)}|{str(self.failure)}|\n"
+
+    rows: list[Row] = []
+
+    def __init__(self, table_file_path: str) -> None:
+        if os.path.exists(table_file_path):
+            with open(table_file_path) as table_file:
+                for i, line in enumerate(table_file):
+                    if i < 2:
+                        continue
+                    stripped_line = line.strip()[1:-1].split("|")
+                    self.rows.append(self.Row().init_from_line(stripped_line))
+
+    def append(
+        self, repo_name: str, repo_link: str, failure_name: str, failure_link: str
+    ) -> None:
+        now = datetime.now().strftime("%Y-%m-%d %H:%M")
+        search_repo = self.Row.Link().init(repo_name, repo_link)
+        row_found = False
+        for i, row in enumerate(self.rows[:]):
+            if row.repository == search_repo:
+                row_found = True
+                if failure_name == "":
+                    self.rows.remove(row)
+                else:
+                    self.rows[i].date = now
+                    self.rows[i].failure = self.Row.Link().init(
+                        failure_name, failure_link
+                    )
+                break
+        if not row_found and failure_name != "":
+            self.rows.append(
+                self.Row().init(now, repo_name, repo_link, failure_name, failure_link)
+            )
+
+    def append_from_score_file(
+        self, score_file_path: str, repo_name: str, repo_link: str
+    ) -> None:
+        with open(score_file_path) as json_file:
+            scorefile: dict[str, Any] = json.load(json_file)
+        failed_name = ""
+        fail_found = False
+        for testrecord in scorefile["testrecords"]:
+            if fail_found:
+                break
+            testname = testrecord["testname"]
+            for result in testrecord["stageresults"]:
+                name = result["name"]
+                if result["force_quit"] == True:
+                    failed_name = f"{testname}/{name}"
+                    fail_found = True
+
+        self.append(repo_name, repo_link, failed_name, "")
+        # TODO: What is failed_link?
```

Comment:
There was supposed to be a link to the failed stage. Previously each stage is in a separate drone step with its own url, so it was implementable. Now all stages are not separated, so I don't think there can be a link.